src/backend/InvenTree/order/models.py

```diff
@@ -599,13 +630,181 @@ def add_line_item(
+                permitted = True
+            elif hasattr(self.project_code, 'responsible'):
+                if self.project_code.responsible.is_user_allowed(approver):
+                    permitted = True
+
+            if permitted:
+                self.approved_by = approver
+                self.approved_date = InvenTree.helpers.current_date()
+                self.reject_reason = ''
+                self._action_ready()
+                trigger_event('purchaseorder.approved')
+                notify_users(
+                    [self.created_by, self.responsible],
+                    self,
+                    PurchaseOrder,
+                    InvenTreeNotificationBodies.PurchaseOrderApproved,
+                    exclude=approver,
+                    delta=timedelta(seconds=60),
+                )
+            else:
+                raise PermissionDenied()
+
+        else:
+            self._action_ready()
+
+    def _action_ready(self, *args, **kwargs):
+        """Marks the PurchaseOrder as READY.
+
+        Order must currently be PENDING or IN_APPROVAL
+        """
+        if not self.requires_approval and not self.can_be_ready:
+            raise ValidationError(_('Ready state setting not active'))
+
+        if self.is_pending or (self.is_pending_approval and self.approved_by):
+            self.status = PurchaseOrderStatus.READY.value
+            self.save()
+
+            user = kwargs.pop('user', None)
+
+            purchaser_group_name = getSetting('PURCHASE_ORDER_PURCHASER_GROUP')
+            if purchaser_group_name:
+                purchasers_group = Group.objects.filter(
+                    name=purchaser_group_name
+                ).first()
+                targets = [purchasers_group, self.created_by]
+                if self.responsible:
+                    targets.append(self.responsible)
+                exclude = []
+                user and exclude.append(user)
+                self.approved_by and exclude.extend([
+                    self.approved_by,
+                    self.created_by,
+                    self.responsible,
+                ])
+
+                notify_users(
+                    targets,
+                    self,
+                    PurchaseOrder,
+                    InvenTreeNotificationBodies.PurchaseOrderReady,
+                    delta=timedelta(seconds=60),
+                    exclude=exclude,
+                )
+
     def _action_place(self, *args, **kwargs):
         """Marks the PurchaseOrder as PLACED.
 
         Order must be currently PENDING.
         """
-        if self.is_pending:
+        purchaser_group = getSetting('PURCHASE_ORDER_PURCHASER_GROUP')
+        if self.is_pending or self.is_ready:
+            user = kwargs.pop('user', None)
+            if (
+                purchaser_group
+                and not user.groups.filter(name=purchaser_group).exists()
+            ):
+                raise PermissionDenied()
```

Comment:
Can we add a custom message to the `PermissionDenied` error?

Reply:
Sure thing